pandas/tests/frame/test_subclass.py

```diff
@@ -773,3 +773,31 @@ def test_constructor_with_metadata():
     )
     subset = df[["A", "B"]]
     assert isinstance(subset, MySubclassWithMetadata)
+
+
+class SimpleDataFrameSubClass(DataFrame):
+    """A subclass of DataFrame that does not define a constructor."""
+
+
+class SimpleSeriesSubClass(Series):
+    """A subclass of Series that does not define a constructor."""
+
+
+class TestSubclassWithoutConstructor:
+    def test_copy_df(self):
+        expected = DataFrame({"a": [1, 2, 3]})
+        result = SimpleDataFrameSubClass(expected).copy()
+
+        tm.assert_frame_equal(result, expected)
```

Comment:
Although we have a `check_frame_type` ("Whether to check the DataFrame class is identical.") that defaults to True, that seems to only check that the left frame is an instance of the class of the right frame. So that won't actually catch the issue if `result` was still a subclass instead of a DataFrame.

So in addition I would add an explicit:
```suggestion
        assert type(result) is DataFrame
        tm.assert_frame_equal(result, expected)
```

Reply:
Huh! Weird!

But I think this does explain some confusion I've had about this function. I've added a comment so this doesn't get removed by someone assuming the type check is covered.

```suggestion
        assert type(result) is DataFrame  # assert_frame_equal only checks isinstance(lhs, type(rhs))
        tm.assert_frame_equal(result, expected)
```

Reply:
`assert_series_equal` seems order invariant here

Reply:
We should probably call that a bug in `assert_frame_equal` ..